BJS2-66099-post-system

marsel-mkh opened the pull request:
Протестирована Postman
createDraft:
![image](https://github.com/user-attachments/assets/b6ff1c77-a4a0-4675-80d8-29ce06cc240d)
publishPost:
![image](https://github.com/user-attachments/assets/afe838a8-a879-42e2-9b28-e2a63e68a66f)
updatePost:
![image](https://github.com/user-attachments/assets/c6073171-41c1-4a0a-b3c3-0d17af0a7c81)
softDeletePost:
![image](https://github.com/user-attachments/assets/a36b3f90-ab33-4b93-9b66-8612b8ad96da)
getPost:
![image](https://github.com/user-attachments/assets/46ae94ab-0205-4b29-a962-c9d46db5ad6e)
getUserDraft:
![image](https://github.com/user-attachments/assets/cc3b91c7-2178-4eac-bfb9-1e63d74bbe1e)
getProjectDraft:
![image](https://github.com/user-attachments/assets/d405d475-252b-4f0a-9ead-27b73b57117c)
getAuthorPublishedPost:
![image](https://github.com/user-attachments/assets/733e271b-61fd-4a78-9c2e-09db6d0dae55)
getProjectPublishedPost:
![image](https://github.com/user-attachments/assets/c0d2db50-1c01-46be-aee4-afda0ed4e420)

Commit:
BJS2-66099-post-system

Commit:
BJS2-66099-post-system
FIX: add spaces

Commit:
BJS2-66099-post-system
added test, new exception in GlobalExceptionHandler

Commit:
BJS2-66099-post-system
fix spaces

Commit:
BJS2-66099-post-system
added GlobalExceptionHandlerTest with Unit Test

Commit:
BJS2-66099-post-system
отредактированы методы под таску Гульназ:
добавлено поле likeCounter
метод getAuthorPublishedPost возвращает дто со списком лайков
getProjectPublishedPost возвращает дто со списком лайков
тесты адаптированы под измнения
измненен маппер(добавлен подсчет лайков)
Поле authorId теперь может быть null

gulnaz21 commented on src/main/java/faang/school/postservice/validation/PostValidator.java:
```diff
@@ -0,0 +1,90 @@
+            throw new DataValidationException("Author or Project in the post is not found");
+        }
+
+        if(authorId != null){
```
Что если оба не null? В таком случае ведь у нас как автор будет указан и пользователь и проект, но по условию нужно, чтобы было что-то одно

marsel-mkh replied:
По умолчанию автором является автор,если автора нет, то в качестве автора ставится проект

gulnaz21 commented on src/main/java/faang/school/postservice/validation/PostValidator.java:
```diff
@@ -0,0 +1,90 @@
+                throw new EntityNotFoundException("Author does not exist");
+            }
+        } else {
+            if(projectServiceClient.getProject(projectId) == null){
```
можно вынести getUser() и getProject() в отдельные переменные для лучшей читаемости кода

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+    public PostViewDto publishPost(long postId) {
+        Post publishPost = postRepository.findById(postId).orElseThrow(() -> {
+            log.error("Post with ID {} not found", postId);
+            return new EntityNotFoundException("Post not found with id: " + postId);
```
Думаю лучше использовать String.format() вместо обычной конкатенации

gulnaz21 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,212 @@
+                                                  @PathVariable long postId) {
+        log.info("Received request to update post with ID: {}", postId);
+
+        postValidator.validateAuthor(postUpdateDto, postId);
```
Я думаю, что эта проверка должна быть в PostService 🤔

marsel-mkh replied:
убрал автора из дто впринципе, так что и проверка больше не нужна

gulnaz21 replied:
супер

gulnaz21 commented on src/main/java/faang/school/postservice/validation/PostValidator.java:
```diff
@@ -0,0 +1,90 @@
+        Long oldAuthorId = oldPost.getAuthorId();
+        Long newAuthorId = postUpdateDto.getAuthorId();
+
+        if (!newAuthorId.equals(oldAuthorId)) {
```
что если автором поста выступает проект? Ведь может быть ситуация, когда authorId = null, а projectId передали измененный

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    @Transactional
+    public PostViewDto updatePost(@NotNull PostUpdateDto postUpdateDto, long postId) {
+        Post oldPost = postRepository.findById(postId).orElseThrow(() -> {
```
Может быть есть смысл вынести данный код в отдельный метод, чтобы не дублировать его

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+    public PostViewDto softDeletePost(long postId) {
+        Post post = postRepository.findById(postId).orElseThrow(() -> {
+            log.error("Post with ID {} not found", postId);
+            return new EntityNotFoundException("Post not found with id: " + postId);
```
здесь тоже лучше String.format() использовать

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    public List<PostViewDto> getUserDraft(long userId) {
+        return postRepository.findByAuthorId(userId).stream()
+                .filter(post-> !(post.isDeleted()))
```
Думаю, может стоит еще добавить фильтрацию на !published?

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    public List<PostViewDto> getProjectDraft(long projectId) {
+        return postRepository.findByProjectId(projectId).stream()
+                .filter(post-> !(post.isDeleted()))
```
здесь тоже можно добавить фильтрацию на !published

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    public List<PostViewDto> getAuthorPublishedPost(long userId) {
+        return postRepository.findByAuthorIdWithLikes(userId).stream()
+                .filter(post-> !(post.isDeleted()))
```
здесь нужно добавить фильтрацию на published = true

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+     */
+    public List<PostViewDto> getProjectPublishedPost(long projectId) {
+        return postRepository.findByProjectIdWithLikes(projectId).stream()
+                .filter(post-> !(post.isDeleted()))
```
и здесь фильтрацию на published = true

gulnaz21 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,212 @@
+     * @return ResponseEntity со списком черновиков в формате PostViewDto.
+     */
+    @GetMapping("/user/{userId}/draft")
+    public ResponseEntity<List<PostViewDto>> getUserDraft(@PathVariable long userId) {
```
Может назвать этот метод getUserDraftPosts или getUserPostDrafts, ну либо getUserDrafts, чтобы показать, что возвращается список черновиков

marsel-mkh replied:
Черновики и посты разные вещи, про множественное число согласен

gulnaz21 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,212 @@
+     * @return ResponseEntity со списком черновиков в формате PostViewDto.
+     */
+    @GetMapping("/project/{projectId}/draft")
+    public ResponseEntity<List<PostViewDto>> getProjectDraft(@PathVariable long projectId) {
```
здесь также как и с названием getUserDraft()

gulnaz21 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,212 @@
+     * @return ResponseEntity со списком опубликованных постов в формате PostViewDto.
+     */
+    @GetMapping("/user/{userId}/published-post")
+    public ResponseEntity<List<PostViewDto>> getAuthorPublishedPost(@PathVariable long userId) {
```
getAuthorPublishedPosts, чтобы уточнить что список постов возвращается, а не 1 пост

gulnaz21 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,212 @@
+     * @return ResponseEntity со списком опубликованных постов в формате PostViewDto.
+     */
+    @GetMapping("/project/{projectId}/published-post")
+    public ResponseEntity<List<PostViewDto>> getProjectPublishedPost(@PathVariable long projectId) {
```
getProjectPublishedPosts

gulnaz21 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,212 @@
+
+        log.info("Post get successfully: {}", postViewDto);
+        return ResponseEntity.ok(postViewDto);
+
```
лишний пробел

gulnaz21 commented on src/main/java/faang/school/postservice/dto/post/PostCreateDto.java:
```diff
@@ -0,0 +1,38 @@
+    private boolean published;
+    private LocalDateTime publishedAt;
+    private LocalDateTime scheduledAt;
+
```
пробел лишний

gulnaz21 commented on src/main/java/faang/school/postservice/dto/post/PostCreateDto.java:
```diff
@@ -0,0 +1,38 @@
+public class PostCreateDto {
+    @NotBlank(message = "Content is not blank")
+    private String content;
+    private Long authorId;
```
Может быть стоит заменить authorId на userId? authorId немного путает, так как автором может выступать не только пользователь, но еще и проект. Что думаешь?

marsel-mkh replied:
userId тоже не совсем понятно, что за юзер и кто он? поэтому думаю лучше оставить authorId

gulnaz21 replied:
просто так у тебя authorId как будто отвечает за автора, а projectId как будто какая-то привязка к проекту какому-то, давай может тогда назовем userAuthorId и projectAuthorId?

marsel-mkh replied:
Тогда в другом контексте может быть непонятно - findByProjectId(projectAuthorId) звучит как то неправильно

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+            return new EntityNotFoundException("Post not found with id: " + postId);
+        });
+
+        if(publishPost.isPublished()){
```
сделай ctrl+alt+l на файлике

gulnaz21 commented on src/main/java/faang/school/postservice/dto/post/PostCreateDto.java:
```diff
@@ -0,0 +1,38 @@
+    private Long projectId;
+    private List<Long> resourceIds;
+    private boolean published;
+    private LocalDateTime publishedAt;
```
я вот думаю, а нужно ли тебе здесь это поле, ведь при создании поста он у тебя сначала создается в состоянии черновика, а публикуешь ты его уже в другом методе, не используя PostCreateDto. Можно оставить конечно, чтобы запоминать дату создания самого черновика, но думаю, это поле не совсем для этого

gulnaz21 commented on src/main/java/faang/school/postservice/dto/post/PostUpdateDto.java:
```diff
@@ -0,0 +1,26 @@
+    private String content;
+    @NotNull(message = "Author can not be null")
+    private Long authorId;
+}
```
можно, наверное, добавить еще resourceIds и scheduledAt

gulnaz21 commented on src/main/java/faang/school/postservice/dto/post/PostUpdateDto.java:
```diff
@@ -0,0 +1,26 @@
+    @NotBlank(message = "Content is not blank")
+    private String content;
+    @NotNull(message = "Author can not be null")
+    private Long authorId;
```
может вообще стоит убрать это поле, раз автора поста изменить или удалить нельзя? Тогда и проверки на изменение автора не нужны будут. Что думаешь?

gulnaz21 commented on src/main/java/faang/school/postservice/mapper/PostMapper.java:
```diff
@@ -0,0 +1,28 @@
+    default long getLikeCounts(List<Like> likes){
+        return likes.size();
+    }
+
```
пробел:)

gulnaz21 commented on src/main/java/faang/school/postservice/mapper/PostMapper.java:
```diff
@@ -0,0 +1,28 @@
+@Mapper(componentModel = "spring")
+public interface PostMapper {
+
+    Post createDtoToEntity(PostCreateDto postCreateDto);
```
у тебя resourceIds в resources некорректно замаппятся

gulnaz21 commented on src/main/java/faang/school/postservice/mapper/PostMapper.java:
```diff
@@ -0,0 +1,28 @@
+    Post createDtoToEntity(PostCreateDto postCreateDto);
+
+    @Mapping(source = "likes", target = "likeCounter")
+    PostViewDto toViewDto(Post post);
```
а здесь commentIds, resourceIds и adId. Нужно будет также создать дефолтные методы для commentIds и resourceIds, где принимаешь, например, `List<Comment>`, а возвращаешь `List<Long>`, для adId можно просто использовать "ad.id"

gulnaz21 commented on src/main/java/faang/school/postservice/model/Post.java:
```diff
@@ -31,53 +53,102 @@
     @Column(name = "content", nullable = false, length = 4096)
     private String content;
 
-    @Column(name = "author_id", nullable = false)
```
Правильно, оно и не надо:)

gulnaz21 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,202 @@
+                .sorted(Comparator.comparing(Post::getCreatedAt).reversed())
+                .map(postMapper::toViewDto)
+                .toList();
+
```
пробел 🥹

gulnaz21 commented on src/main/java/faang/school/postservice/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,86 @@
+ * Этот класс перехватывает исключения, возникающие в контроллерах, и возвращает соответствующие HTTP-ответы.
+ */
+@Slf4j
+@ControllerAdvice
```
можно использовать @RestControllerAdvice, раз мы везде используем @RestController

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+    @DisplayName("Проверка успешной валидации, когда передан valid PostCreateDto с authorId")
+    @Test
+    public void givenValidAuthorIdWhenValidateAuthorAndProjectThenSuccessfulValidation() {
+        postCreateDto = new PostCreateDto();
```
можно вынести в `@BeforeEach`

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+        postCreateDto.setAuthorId(VALID_AUTHOR_ID);
+        userDto = new UserDto(1L,"name","mail");
+
+        Mockito.when(userServiceClient.getUser(postCreateDto.getAuthorId())).thenReturn(userDto);
```
давай postCreateDto.getAuthorId() вынесем в отдельную переменную, либо просто будем использовать VALID_AUTHOR_ID

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+    @Test
+    @DisplayName("Проверка успешной валидации, когда передан valid PostCreateDto с projectId")
+    public void givenProjectIdWhenValidateAuthorAndProjectThenSuccessfulValidation() {
+        PostCreateDto post = new PostCreateDto();
```
у тебя уже есть созданный PostCreateDto , можно использовать его

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+    @DisplayName("Проверка успешной валидации, когда передан valid PostCreateDto с projectId")
+    public void givenProjectIdWhenValidateAuthorAndProjectThenSuccessfulValidation() {
+        PostCreateDto post = new PostCreateDto();
+        post.setAuthorId(null);
```
если создаешь новый PostCreateDto, то эта строчка будет лишней

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+
+       projectDto = new ProjectDto(1L,"Test Project");
+
+        Mockito.when(projectServiceClient.getProject(1L)).thenReturn(projectDto);
```
давай 1L заменим на VALID_PROJECT_ID

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+    public void givenInvalidAuthorIdAndProjectIdWhenValidateAuthorAndProjectThenDataValidationException() {
+        PostCreateDto post = new PostCreateDto();
+        post.setAuthorId(null);
+        post.setProjectId(null);
```
тут так же, как и выше, можно просто использовать уже созданный postCreateDto. А если еще и вынесешь new PostCreateDto() в `@BeforeEach`, то `post.setAuthorId(null)` и `post.setProjectId(null)` можно будет не прописывать, просто запишешь id там, где это нужно будет

gulnaz21 commented on src/main/java/faang/school/postservice/validation/PostValidator.java:
```diff
@@ -0,0 +1,90 @@
+
+        if (authorId == null && projectId == null) {
+            log.warn("Validation failed: Both author and project are missing in the post");
+            throw new DataValidationException("Author or Project in the post is not found");
```
я думаю можно придумать какое-нибудь другое сообщение об ошибке, "is not found" выглядит так, будто мы не нашли их в бд, хотя у нас просто пришли значения null. Можно вывести что-то типо "у поста не может отсутствовать автор"

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+    public void givenNonExistentAuthorIdWhenValidateAuthorAndProjectThenEntityNotFoundException() {
+        PostCreateDto post = new PostCreateDto();
+        post.setAuthorId(VALID_AUTHOR_ID);
+        post.setProjectId(null);
```
то же, что и выше писала

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+        post.setAuthorId(VALID_AUTHOR_ID);
+        post.setProjectId(null);
+
+        Mockito.when(userServiceClient.getUser(1L)).thenReturn(null);
```
константу можно заменить на VALID_AUTHOR_ID

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+        post.setAuthorId(null);
+        post.setProjectId(VALID_PROJECT_ID);
+
+        Mockito.when(projectServiceClient.getProject(1L)).thenReturn(null);
```
тут так же, как и выше ☝

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostValidatorTest.java:
```diff
@@ -0,0 +1,170 @@
+            postValidator.validateAuthorAndProject(post));
+
+        Assertions.assertEquals("Project does not exist", exception.getMessage());
+    }
```
не забудь добавить проверку, когда и authorId и projectId = null

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+        postViewDto = new PostViewDto();
+        postUpdateDto = new PostUpdateDto();
+        post = new Post();
+
```
пробел

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка успешной публикации поста")
+    public void givenVaLidPostIdWhenPublishPostThenReturnPostViewDto() {
+        long postId = VALID_POST_ID;
+        post.setPublished(false);
```
можно думаю не писать это, по-умолчанию ведь false значение

marsel-mkh replied:
Так явно видно, что не опубликовано

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+
+        Assertions.assertNotNull(postService.updatePost(postUpdateDto, postId));
+
+        Mockito.verify(postMapper, Mockito.times(1)).toViewDto(post);
```
можно тогда еще проверить, что вызвался postMapper.update()

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @Test
+    @DisplayName("Проверка успешного получения поста")
+    public void givenValidPostIdWhenGetPostThenReturnPostViewDto() {
+        long postId =VALID_POST_ID;
```
ctrl+alt+l

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка успешного получения поста")
+    public void givenValidPostIdWhenGetPostThenReturnPostViewDto() {
+        long postId =VALID_POST_ID;
+        post = new Post();
```
эта строчка лишняя, метод setUp() и так вызывает post = new Post()

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка успешного получения черновиков пользователя")
+    public void givenValidUserIdWhenGetUserDraftThenReturnListOfPostViewDto() {
+        long userId = VALID_USER_ID;
+        post.setDeleted(false);
```
думаю лишнее

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка получения пустого списка черновиков пользователя")
+    public void givenInvalidUserIdWhenGetUserDraftThenReturnListOfPostViewDto() {
+        long userId = VALID_USER_ID;
+        post.setDeleted(false);
```
здесь тоже думаю лишнее, по-умолчанию false стоит

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка успешного получения черновиков проекта")
+    public void givenValidProjectIdWhenGetProjectDraftThenReturnListOfPostViewDto() {
+        long projectId = VALID_PROJECT_ID;
+        post.setDeleted(false);
```
думаю, лишнее

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка получения пустого списка черновиков проекта")
+    public void givenInvalidProjectIdWhenGetProjectDraftThenReturnListOfPostViewDto() {
+        long projectId = VALID_PROJECT_ID;
+        post.setDeleted(false);
```
и тут тоже

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка успешного получения опубликованных постов пользователя")
+    public void givenValidUserIdWhenGetAuthorPublishedPostThenReturnListOfPostViewDto() {
+        long userId = VALID_USER_ID;
+        post.setDeleted(false);
```
и здесь

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка получения пустого списка опубликованных постов пользователя")
+    public void givenInvalidUserIdWhenGetAuthorPublishedPostThenReturnListOfPostViewDto() {
+        long userId = VALID_USER_ID;
+        post.setDeleted(false);
```
и тут

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+        post.setDeleted(false);
+        post.setPublishedAt(LocalDateTime.now());
+
+        Mockito.when(postRepository.findByAuthorId(userId)).thenReturn(List.of());
```
здесь наверное findByAuthorIdWithLikes() нужно

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+
+        Mockito.when(postRepository.findByAuthorId(userId)).thenReturn(List.of());
+
+        Assertions.assertNotNull(postService.getUserDraft(userId));
```
а здесь вызывать getAuthorPublishedPost()

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка успешного получения опубликованных постов проекта")
+    public void givenValidProjectIdWhenGetProjectPublishedPostThenReturnListOfPostViewDto() {
+        long projectId = VALID_PROJECT_ID;
+        post.setDeleted(false);
```
лишнее

gulnaz21 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,337 @@
+    @DisplayName("Проверка получения пустого списка опубликованных постов проекта")
+    public void givenInvalidProjectIdWhenGetProjectPublishedPostThenReturnListOfPostViewDto() {
+        long projectId = VALID_PROJECT_ID;
+        post.setDeleted(false);
```
и тут

Commit:
BJS2-66099-post-system
из основного: убрана одна из валидаций, валидация перенесена в сервис, чекстайл

Commit:
BJS2-66099-post-system
добавлен сваггер и аннотации на контроллер и дто

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,268 @@
+    )
+    @PutMapping("/{postId}/publish")
+    public ResponseEntity<PostViewDto> publishPost(
+            @PathVariable @Parameter( description = "id поста",required = true, example = "1") long postId) {
```
id мы обертками делаем, а не примитивами

RinatBeybutov commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,268 @@
+     * @return ResponseEntity с опубликованным постом в формате PostViewDto.
+     */
+    @Operation(
+            description = "Публикует пост по его ID." +
```
покажи как эта надпись будет выглядеть в сваггере

marsel-mkh replied:
![image](https://github.com/user-attachments/assets/003ecf05-fce6-4c39-b097-5295f8aede2b)
 Переносы только для того чтоб читабельнее было. Переделал под summary + descrption для читабельности

marsel-mkh replied:
стало так 
![image](https://github.com/user-attachments/assets/d2a3d7fc-419b-4f00-9303-d8ab1e2f484e)

RinatBeybutov commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,211 @@
+     *
+     * @param postCreateDto DTO с данными для создания поста.
+     * @return {@link PostViewDto} - DTO с данными созданного поста.
+     */
```
если уж добавляешь транзакции, то добавляй на все методы, где взаимодействие с базой происходит

RinatBeybutov commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,211 @@
+     */
+    public List<PostViewDto> getUserDrafts(long userId) {
+        return postRepository.findByAuthorId(userId).stream()
+                .filter(post -> !(post.isDeleted()))
```
для чего post.isDeleted() в скобки поместил?
далее идет такой же код, но без скобок. Давай в одном стиле сделаем

safon4ik10 commented on src/main/java/faang/school/postservice/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,86 @@
+     */
+    @ExceptionHandler(DataValidationException.class)
+    public ResponseEntity<String> handleDataValidationException(DataValidationException dataValidationException) {
+        log.warn("Data validation error: {}", dataValidationException.getMessage(), dataValidationException);
```
Тут у тебя Warn, ниже error, лучше тут бросать error это всё таки exception

safon4ik10 commented on src/main/java/faang/school/postservice/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,86 @@
+     *             <li>HTTP-статус {@link HttpStatus#BAD_REQUEST} (400), указывающий на некорректный запрос
+     */
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<Map<String, String>> handleValidationExceptions(MethodArgumentNotValidException exception) {
```
Добавь логи

safon4ik10 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,268 @@
+                    "Параметры: postId ID поста для мягкого удаления." +
+                    "Возвращает: ResponseEntity с мягко удаленным постом в формате PostViewDto."
+    )
+    @PutMapping("/{postId}/soft-delete")
```
Тут должен быть DeleteMapping потому что ты удаляешь сущность

marsel-mkh replied:
Я думаю можно и DeleteMapping  и PutMapping, потому что конкретно никакой объект мы не удаляем, а просто задаем флаг для удаления

safon4ik10 replied:
Да, чет видать глаз не туда улетел, не заметил что в сервисе просто ставится флаг

safon4ik10 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,268 @@
+
+import java.util.List;
+
+/**
```
Я не  уверен что JavaDoc может быть применим для контроллеров, т.к. это делается чисто для фронта, а фронт разработчик пойдет смотреть твой сваггер

safon4ik10 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,268 @@
+                    "Возвращает ResponseEntity с созданным постом в формате PostViewDto."
+    )
+    @PostMapping
+    public ResponseEntity<PostViewDto> createDraft(@Valid @RequestBody PostCreateDto postCreateDto) {
```
Это чисто ИМХО, но зачем в контроллере логи? У него одна задача - передать/получить информацию

safon4ik10 commented on src/main/java/faang/school/postservice/controller/PostController.java:
```diff
@@ -0,0 +1,268 @@
+                    "Возвращает ResponseEntity с опубликованным постом в формате PostViewDto"
+    )
+    @PutMapping("/{postId}/publish")
+    public ResponseEntity<PostViewDto> publishPost(
```
А что если сюда придет Null?

marsel-mkh replied:
Если клиент попытается отправить запрос с null вместо числа (например /null/publish), Spring даже не вызовет ваш метод контроллера, а вернет ошибку 400 (Bad Request) на этапе парсинга URL, так как null не является валидным значением для типа long.

RinatBeybutov replied:
давай заменим на обертки везде. В сущности entity у тебя там Long

safon4ik10 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,211 @@
+     */
+    private Post getPostEntity(long postId) {
+        return postRepository.findById(postId).orElseThrow(() -> {
+            log.error("Post with ID {} not found", postId);
```
У тебя задвоение логов идет, у тебя тут будет лог и в Handler'e

safon4ik10 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,211 @@
+        Post post = getPostEntity(postId);
+
+        if (post.isDeleted()) {
+            throw new DataValidationException("Post is already deleted");
```
Какой пост уже удален?)

safon4ik10 commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -0,0 +1,211 @@
+     * @throws EntityNotFoundException если пост с указанным ID не найден.
+     */
+    public PostViewDto getPost(long postId) {
+        Post post = getPostEntity(postId);
```
Если ты не делаешь логов, лучше в одну строку сразу return

marsel-mkh replied:
Так более читабельно, чем все в одну строку

safon4ik10 commented on src/test/java/faang/school/postservice/util/GlobalExceptionHandlerTest.java:
```diff
@@ -0,0 +1,75 @@
+
+        response = globalExceptionHandler.handleDataValidationException(exception);
+
+        Assertions.assertEquals(HttpStatus.BAD_REQUEST, response.getStatusCode());
```
Со статичным импортом читается более приятно, имхо)

safon4ik10 commented on src/test/java/faang/school/postservice/util/PostServiceTest.java:
```diff
@@ -0,0 +1,342 @@
+    @Test
+    @DisplayName("Проверка ошибки при публикации несуществующего поста")
+    public void givenNotExistPostIdWhenPublishPostThenReturnEntityNotFoundException() {
+        long postId = INVALID_POST_ID;
```
Не кажется каким то лишним присвоением? Ставь сразу статичную переменную сразу будет понятно, что ID поста валидный

Commit:
BJS2-66099-post-system
Добавил в хэндлер npe и FeignException для таски Гульназ

Commit:
BJS2-66099-post-system
переделал сваггер, добавил @Transactional к методам которые изменят данные

RinatBeybutov commented:
вмержи изменения из мастера и разреши конфликты, после этого смержим

Commit:
Merge branch 'murloc-master-stream9' into feature/BJS2-66099-post-system

# Conflicts:
#	src/main/java/faang/school/postservice/controller/GlobalExceptionHandler.java
#	src/main/java/faang/school/postservice/service/PostService.java

Commit:
BJS2-66099-post-system
fix mapper, sort test in the folders

Commit:
BJS2-66099-post-system
deleted npe test
